design-proposals/klip-11-DQL.md

```diff
@@ -1 +1,745 @@
+      hence, at any point in wall-clock time a TABLE might be in an inconsistent state
+      (we call those the _generation_ of a table, i.e., for each table version, there might be multiple generation and only the last generation is consistent).
+  * a TABLE can be represented as (versioned) state or changelog – we consider the physical representation as an implementation detail
+
+
+
+## What is in scope
+
+The focus is to remove `EMIT CHANGES` and introduce new syntax to support different types of queries over streams and tables.
+We want to exploit the stream-table duality and to transform STREAMS and TABLES into each other seamlessly.
+Furthermore, the query language should express the difference between TABLES and MATERIALIZED VIEW explicitly.
+
+## What is not in scope
+
+This KLIP proposes a redesign of the query language but does not propose to implement all possible features right away.
+The intent is to make the language "future proof" such that new feature can be added seamlessly without breaking the language.
+In particular, we should not implement (cf. "Design" section below):
+
+* re-introduce `CREATE TABLE` statement to create RDBMS style tables that allow insert/update/delete statements
+* allow for complex queries over table state, e.g., aggregation and joins
+* we might not need to implement the new `AS OF <timestamp>` clause in the first iteration
+* we might not need to support terminating queries over streams right away
+
+We limit this KLIP to not introduce too much syntactic sugar to express queries.
+This may make some queries unnecessary "complex/clumsy" but the goal is to illustrate to underlying principles of the language.
+We can add syntactic sugar after we agree on the basic design of the language.
+
+
+
+## Value/Return
+
+The "Dual Query Language" should be easier to understand and easier to use.
+It will allow to support nested queries seamlessly and unifies query semantics (persistent vs. transient queries).
+Furthermore, the language is designed to be "future proof", i.e., it considers potential future features and thus allows
+to extend the capabilities of KSQL without the need to change the language in a breaking way.
+
+
+
+## Public APIS
+
+We propose the following language changes:
+
+* introduce `TABLE(<stream>)` operator: takes a STREAM as input, and changes its semantic interpretation from facts to updates; it returns a TABLE
+  * note: the returned TABLE is not an RDBMS style table, but rather a MV
+  * note: it's not specified if the returned TABLE is materialized into a state store or represented as a changelog
+    (it's up the the KSQL optimizer or KS runtime to make a materialization decision)
+  * the `TABLE()` operator consumes it's input STREAM "from beginning"
+  * we could extend `TABLE()` to also accept a TABLE as input if it makes the language smoother;
+    it's obvious that `TABLE(<table>)` would be idempotent, i.e., `TABLE(t) = t`
```

Comment:
you mean an identity function?

Reply:
Ack. Identity is the right term (the identify function is of course idempotent itself...)